plugin/vm_processor.go

```diff
@@ -0,0 +1,128 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/compute/mgmt/compute"
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/resources/mgmt/resources"
+	"golang.org/x/sync/errgroup"
+)
+
+// vmJob represents a single VM processing job
+type vmJob struct {
+	resource      resources.GenericResourceExpanded
+	resourceGroup string
+	name          string
+}
+
+func (p *AzurePlugin) processStandardVMs(ctx context.Context, resourceInfos []resources.GenericResourceExpanded,
+	clients *azureClients, vmToNetworkMap map[string]networkInfo) error {
+
+	// Create error group for worker management
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Create job channel
+	jobs := make(chan vmJob)
+
+	// Start workers
+	workerCount := constDefaultWorkers
+	if len(resourceInfos) < workerCount {
+		workerCount = len(resourceInfos)
+	}
+
+	// Result handling
+	var mu sync.Mutex
+
+	// Start worker pool
+	for i := 0; i < workerCount; i++ {
+		g.Go(func() error {
+			for job := range jobs {
+				// Process single VM
+				vm, err := clients.vmClient.Get(ctx, job.resourceGroup, job.name, "")
+				if err != nil {
+					return fmt.Errorf("error fetching vm with id %q: %w", *job.resource.ID, err)
+				}
+
+				// Only process running VMs
+				running, err := isVMRunning(ctx, clients.vmClient, job.resourceGroup, job.name)
+				if err != nil {
+					return err
+				}
+				if !running {
+					continue
+				}
+
+				var netInfo networkInfo
+				if err := processVMNetworkInterfaces(ctx, vm, clients, job.resourceGroup, &netInfo); err != nil {
+					return err
+				}
+
+				// Thread-safe map update - scope the lock to just the map update
+				func() {
+					mu.Lock()
+					defer mu.Unlock()
+					vmToNetworkMap[*job.resource.ID] = netInfo
+				}()
+			}
+			return nil
+		})
+	}
+
+	// Send jobs to workers
+	go func() {
+		defer close(jobs)
+		for _, res := range resourceInfos {
+			// Check context cancellation
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			resourceGroup, name, err := splitId(*res.ID, constMsComputeService, constVirtualMachinesResource)
+			if err != nil {
+				// Can't send error through channel, so we'll skip this item
+				fmt.Printf("Warning: error splitting vm id %q: %v\n", *res.ID, err)
+				continue
+			}
```

Comment:
This processing should be very cheap, why don't we move it up to before the goroutines so we can error if it fails?

Reply:
Happy to change it, but my thinking here was to just process what we could, rather than duck out if say 1 out of 1000 VMs couldn't be parsed successfully. I guess what is appropriate depends on the UX behaviour we'd rather drive - All or nothing verses as much as is practical?

Reply:
I think we should error and return the failure to the user in this case. The alternative is that we silently fail for a number of resources, which I think is worse. What do you think?

Reply:
Well, I'd say it depends on how you handle the failure scenario. The sweet spot for me would be process what we can and warn on failed resources, but not sure how that'd be handled in the current code paths - or how that actually propagates up to the user at all - I'm not sure how the plugin passes information back to the Boundary control plane.